File: src/main/java/net/notjustanna/audio/AudioInput.java
package net.notjustanna.audio;

import io.micronaut.core.annotation.NonNull;

import javax.sound.sampled.Mixer;
import java.util.*;

public record AudioInput(@NonNull String name, @NonNull String device, @NonNull Set<Mixer.Info> mixers) {
    public String fullName() {
        return name + " (" + device + ")";
    }

    public static class Builder {
        private String name = "";
        private String device = "";
        private final Set<Mixer.Info> mixers = new HashSet<>();

        public String name() {
            return name;
        }

        public Set<Mixer.Info> mixers() {
            return mixers;
        }

        public Builder name(String name) {
            this.name = name;
            return this;
        }

        public Builder device(String device) {
            this.device = device;
            return this;
        }

        public Builder addMixer(Mixer.Info mixer) {
            this.mixers.add(mixer);
            return this;
        }

        public AudioInput build() {
            return new AudioInput(name, device, Set.copyOf(mixers));
        }
    }
}

File: src/main/kotlin/net/notjustanna/audio/system/AudioInput.kt
package net.notjustanna.audio.system

import net.notjustanna.audio.native.MixerInfo

data class AudioInput(
    val name: String,
    val device: String,
    val mixers: Set<MixerInfo>
) {
    val fullName = "$name ($device)"

    class Builder {
        public var name: String = ""
            private set
        public var device: String = ""
            private set
        public var mixers: MutableSet<MixerInfo> = mutableSetOf()
            private set

        fun name(name: String) = apply { this.name = name }

        fun device(device: String) = apply { this.device = device }

        fun mixers(mixers: Collection<MixerInfo>) = apply { this.mixers = mixers.toMutableSet() }
        fun addMixer(mixer: MixerInfo) = apply { mixers.add(mixer) }
        fun addMixers(mixers: Collection<MixerInfo>) = apply { this.mixers.addAll(mixers) }
        fun addMixers(mixers: Array<MixerInfo>) = apply { this.mixers.addAll(mixers) }

        fun build() = AudioInput(name, device, mixers)
    }
}